llama_index/vector_stores/elasticsearch.py

```diff
@@ -0,0 +1,468 @@
+)
+from llama_index.vector_stores.utils import metadata_dict_to_node, node_to_metadata_dict
+
+logger = getLogger(__name__)
+
+DISTANCE_STRATEGIES = Literal[
+    "COSINE",
+    "DOT_PRODUCT",
+    "EUCLIDEAN_DISTANCE",
+]
+
+
+def _get_elasticsearch_client(
+    *,
+    es_url: Optional[str] = None,
+    cloud_id: Optional[str] = None,
+    api_key: Optional[str] = None,
+    username: Optional[str] = None,
+    password: Optional[str] = None,
+) -> Any:
+    """Get Elasticsearch client.
+
+    Args:
+        es_url: Elasticsearch URL.
+        cloud_id: Elasticsearch cloud ID.
+        api_key: Elasticsearch API key.
+        username: Elasticsearch username.
+        password: Elasticsearch password.
+
+    Returns:
+        Elasticsearch client.
+
+    Raises:
+        ConnectionError: If Elasticsearch client cannot connect to Elasticsearch.
+    """
+
+    try:
+        import elasticsearch
+    except ImportError:
+        raise ImportError(
+            "Could not import elasticsearch python package. "
+            "Please install it with `pip install elasticsearch`."
+        )
+
+    if es_url and cloud_id:
+        raise ValueError(
+            "Both es_url and cloud_id are defined. Please provide only one."
+        )
+
+    if es_url and cloud_id:
+        raise ValueError(
+            "Both es_url and cloud_id are defined. Please provide only one."
+        )
+
+    connection_params: Dict[str, Any] = {}
+
+    if es_url:
+        connection_params["hosts"] = [es_url]
+    elif cloud_id:
+        connection_params["cloud_id"] = cloud_id
+    else:
+        raise ValueError("Please provide either elasticsearch_url or cloud_id.")
+
+    if api_key:
+        connection_params["api_key"] = api_key
+    elif username and password:
+        connection_params["basic_auth"] = (username, password)
+
+    es_client = elasticsearch.Elasticsearch(**connection_params)
+    try:
+        es_client.info()
+    except Exception as e:
+        logger.error(f"Error connecting to Elasticsearch: {e}")
+        raise e
+
+    return es_client
+
+
+def _to_elasticsearch_filter(standard_filters: MetadataFilters) -> Dict[str, Any]:
+    """Convert standard filters to Elasticsearch filter.
+
+    Args:
+        standard_filters: Standard Llama-index filters.
+
+    Returns:
+        Elasticsearch filter.
+    """
+    if len(standard_filters.filters) == 1:
+        filter = standard_filters.filters[0]
+        return {
+            "term": {
+                f"metadata.{filter.key}.keyword": {
+                    "value": filter.value,
+                }
+            }
+        }
+    else:
+        operands = []
+        for filter in standard_filters.filters:
+            operands.append(
+                {
+                    "term": {
+                        f"metadata.{filter.key}.keyword": {
+                            "value": filter.value,
+                        }
+                    }
+                }
+            )
```

Comment:
If you send many filters with the same key (e.g: document file names) this will generate an AND instead of an OR. What about grouping by key before this and using "terms" instead of "term" ?

Reply:
I think its an outer and , but inner OR.

This is the modification I did locally:

```
    unique_keys = set([filter.key for filter in standard_filters.filters])

    for key in unique_keys:
        values = [filter.value for filter in standard_filters.filters if filter.key == key]
        operands.append(
            {
                "terms": {
                    f"metadata.{key}.keyword": values
                }
            }
        )

    return {"bool": {"must": operands}}
```

So will do an inner OR , outer AND